test/plugin-app.js

```diff
@@ -8,13 +9,57 @@ t.test('Plugin app', async t => {
   if (app.mode === 'development') app.log.level = 'debug';
 
   app.static.publicPaths.push(Path.currentFile().sibling('support', 'js', 'static-app', 'public').toString());
+  const channel = 'concurrent-helpers-test';
+  app.models.events = new EventEmitter();
 
   app.plugin(mixedPlugin);
 
   app.get('/tag_helpers', ctx => ctx.render({inline: tagHelperPlugin}));
   app.get('/form_helpers', ctx => ctx.render({inline: formTagHelpers}));
 
   app.get('/helper', ctx => ctx.render({text: ctx.testHelper('test')}));
+  app.get('/nested/helper', ctx => ctx.render({text: ctx.nested.testHelper('test')}));
+  app.get('/nested/helper_with_stash/:test', ctx => ctx.render({text: ctx.nested.testStashHelper('test')}));
+  app.get('/nested/other_helper_with_stash/:test', ctx => ctx.render({text: ctx.other.otherStashHelper('test')}));
+  app.get('/nested/innexistent_helper_with_stash/:test', ctx => ctx.render({text: ctx.other.testStashHelper('test')}));
+  app.get('/concurrent_helpers/first', async ctx => {
+    const requestId = ctx.req.requestId;
+    const nested = ctx.nested; // save nested getter
+    let sync = false;
+    const listener = msg => {
+      sync ||= msg === 'continue';
+    };
+    ctx.models.events.on(channel, listener);
+    while (!sync) {
+      await new Promise(r => setTimeout(r, 1));
+      await ctx.models.events.emit(channel, 'start');
+    }
+    const helperFinalId = nested.requestId();
+    if (helperFinalId === requestId) return await ctx.render({text: `Ok, requestId = ${requestId} everywhere`});
+    await ctx.render({
+      text: `Not Ok, requesId: ${requestId} , helper read: ${helperFinalId}`,
+      status: 500
+    });
+  });
+  app.get('/concurrent_helpers/second', async ctx => {
+    const requestId = ctx.req.requestId;
+    const nested = ctx.nested; // save nested getter
+    let sync = false;
+    const listener = msg => {
+      sync ||= msg === 'start';
+    };
+    ctx.models.events.on(channel, listener);
+    while (!sync) {
+      await new Promise(r => setTimeout(r, 1));
+    }
+    await ctx.models.events.emit(channel, 'continue');
+    const helperFinalId = nested.requestId();
+    if (helperFinalId === requestId) return await ctx.render({text: `Ok, requestId = ${requestId} everywhere`});
+    await ctx.render({
+      text: `Not Ok, requesId: ${requestId} , helper read: ${helperFinalId}`,
+      status: 500
+    });
+  });
```

Comment:
This test case looks way too complicated and probably prone to race conditions.

Reply:
Should be possible to just make two [mock contexts](https://github.com/mojolicious/mojo.js/blob/e95755f6da4b3c6a519c1d4a638130a1816527b4/test/app.js#L933) and use those to trigger the problem with very little code.

Reply:
done in last commit.
Previous to this one, there is a cache implemented inside the getter, so the resulting returned object is calculated only when the first get() happens.

About the Proxy implementation, I probably would need to proxy the ctx object itself, as according to the intents I was able to do so far it is the only way to actually receive the proper `this` (as target) in the get(target, prop, receiver) method from the proxy definition.

Is that the suggested approach? I mean to proxy the context object and not the ctx.nested object.

Reply:
I have not suggested a specific approach.